### DISCUSSION TITLE
Setting the current surface's theme

### DISCUSSION THREAD

**Discussion by devshane** [ghostty-org/ghostty#3457] [Feature Requests, Ideas] [answered]
I'm hoping to clarify a comment discussing themes and new windows: https://github.com/ghostty-org/ghostty/discussions/2363#discussioncomment-10824717

My use case is to set the theme in the current session before I ssh to a staging or production server and then revert it when the connection is closed. Other terms support this workflow with escape codes or shell integration.

Is this possible now? If not, is it a planned feature?

**Comment by jscheel** [comment]
This is a huge one for me. I have a script called `prodmode` that I use in other terminal emulators. It sets a bunch of env vars, asdf shims, and the session theme. This would be awesome to be able to do in ghostty.

**Comment by rhodes-b** [comment]
There is no way to programatically control ghostty right now, or have a theme in a specific situation besides it detecting system light mode and dark mode changes. Ghostty plans on having support to be programmatically controlled in the future refer to #2353 for more details on that

(besides editing an reloading the config)

**Reply by jscheel** [reply]
I think the specifics about this question have more to do with per-session configs, which is adjacent to the discussion about an api. But yeah, the other thread is definitely the place to follow for the dev of an api.

**Reply by rhodes-b** [reply]
> I think the specifics about this question have more to do with per-session configs, which is adjacent to the discussion about an api. But yeah, the other thread is definitely the place to follow for the dev of an api.

Which also dont exist like listed in that discussion op linked. 

> Other terms support this workflow with escape codes 

Kitty for example lets remote hosts change themes so it is related in that case

**Comment by 00-kat** [comment, marked as answer]
There's an escape sequence to set the theme ([OSC 4, 5, 6, 10-19, 104, 105, 106, 110-119][xterm], and there's also [Kitty's extension which Ghostty also supports][kitty].  I recommend using [Kitty's extension][kitty] if your copy has #5770 as it's easier to use), which you could probably use as a workaround in your shell configuration or something.  But as Bruce said, this is a duplicate of #2353.

As an example of using **Kitty's escape sequence**:
```sh
printf '\033]21;foreground=#FFFFFF;background=#000000;3=#5F5E2A;6=#227685\a'
```

And an example of using **XTerm's smattering of OSC sequences**:
```sh
printf '\e]10;#FFFFFF\a'
printf '\e]11;#000000\a'
printf '\e]4;3=#5F5E2A;6=#227685\a'
```
Which can also be sent in a single `printf`:
```sh
printf '\e]10;#FFFFFF\a\e]11;#000000\a\e]4;3=#5F5E2A;6=#227685\a'
```
All the examples above are equivalent to the following configuration in the Ghostty config format:
```ini
foreground = #FFFFFF
background = #000000
palette = 3=#5F5E2A
palette = 6=#227685
```

[xterm]: https://pod.tst.eu/http://cvs.schmorp.de/rxvt-unicode/doc/rxvt.7.pod#XTerm_Operating_System_Commands
[kitty]: https://sw.kovidgoyal.net/kitty/color-stack/#setting-and-querying-colors

**Reply by devshane** [reply]
When I paste that `printf` example at a zsh (or bash) prompt, the colors are changed correctly but I see control characters printed and "21" is added to the keyboard buffer as if I'd typed it:

```
bash-3.2$ printf '\033]21;foreground=#FFFFFF;background=#000000;3=#5F5E2A;6=#227685\e\\'
^[]21^[\bash-3.2$ 21
```

It doesn't seem like ghostty is "eating" the kitty escape sequences. When I use OSC sequences, say for just the background color (11), it operates as I'd expect. For example, this changes the background to bright green without any outputted characters or keyboard buffer input:

```
bash-3.2$ printf '\033]11;#00ff00\007'
bash-3.2$
```

The alternative ST sequence also works:

```
bash-3.2$ printf '\033]11;#0000ff\e\\'
bash-3.2$
```

Thanks for the help, it's appreciated! This is very close to what I need atm.

**Reply by 00-kat** [reply]
Hmm, odd.  I had tried out my `printf` in Fish, but in Bash I can see odd behaviour when I use it too.

It looks like Ghostty is sending the following in response to that OSC: `\033]21\033\\` (i.e. `OSC 21 ST`), which Kitty doesn't do.  The spec doesn't specify whether responding with nothing in the presence of no query is allowed, and it looks like Ghostty does; [I'll ask around] but for now I'll update my comment to use the XTerm set of sequences.

[I'll ask around]: https://discord.com/channels/1005603569187160125/1005603569711452192/1339835327933714463

**Reply by 00-kat** [reply]
Will be fixed with #5770.

**Reply by andelink** [reply]
Here's a snippet I use to set the theme for just the current surface, reading the `palette = ...` configs from stdin, building off the great reference from @00-kat:

```shell
ghostty +show-config \
  | perl -ne 's/^palette *= *//, s/-/_/g, s/ = /=/, print if /= ?#/' \
  | xargs -n1 printf $'\e]21;%s\e\\' >> "${TTY}"
```

**Reply by Juts2** [reply]
Is there a clever way to revert the color change when exiting ssh?

**Reply by 00-kat** [reply]
https://sw.kovidgoyal.net/kitty/color-stack/#setting-color-values
> To reset a color to its default value (i.e. the value it would have if it was never set) the client program should send just the key name with no `=` and no value.

I would probably do something like:
```sh
ssh() {
    env ssh "$@"
    printf '\033]21;foreground;background;0;1;2;3;4;5;6;7;8;9;10;11;12;13;14;15\033\\'
}
```

**Reply by Juts2** [reply]
> https://sw.kovidgoyal.net/kitty/color-stack/#setting-color-values
> 
> > To reset a color to its default value (i.e. the value it would have if it was never set) the client program should send just the key name with no `=` and no value.
> 
> I would probably do something like:
> 
> ```shell
> ssh() {
>     env ssh "$@"
>     printf '\033]21;foreground;background;0;1;2;3;4;5;6;7;8;9;10;11;12;13;14;15\033\\'
> }
> ```

Thanks for that, I was working on coming up with something and ended up adding this to my fish config to set/revert the colors. I think yours is cleaner.


```
if set -q SSH_TTY
    printf '\033]21;foreground=#FFFFFF;background=#220000;3=#5F5E2A;6=#227685\a'
end

function on_exit --on-event fish_exit
    #printf '\033]104; \033]105; \033]110; \033]111; \033]112; \033]113; \033]114;\a'
    printf '\033]104;\033]105;\033]110;\033]111;\033]112;\033]114;\a'
end
```
